applications/OptimizationApplication/python_scripts/utilities/response_function_implementor.py

```diff
@@ -0,0 +1,194 @@
+from abc import ABC
+from abc import abstractmethod
+
+import KratosMultiphysics as Kratos
+from KratosMultiphysics.OptimizationApplication.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.responses.response_function import ResponseFunction
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import ContainerVariableDataHolderUnion
+
+class ResponseFunctionImplementor(ABC):
+    def __init__(self, response_name: str, optimization_info: OptimizationInfo):
+        self.__response_function: ResponseFunction = optimization_info.GetOptimizationProcess(ResponseFunction, response_name)
+        self.__name = response_name
+
+        self.__initial_response_value = None
+        self._optimization_info = optimization_info
+        self._key_prefix = f"problem_data/response_data/{self.__response_function.GetModelPart().FullName()}/{self.GetName()}"
+
+    def GetName(self) -> str:
+        return self.__name
+
+    def GetValue(self, solution_step_index: int = 0) -> float:
+        key = f"{self._key_prefix}/value"
+        if solution_step_index == 0 and not self._optimization_info.HasValue(key):
+            self._optimization_info.SetValue(key, self.__response_function.CalculateValue())
+        return self._optimization_info.GetValue(key, solution_step_index)
+
+    def GetSensitivity(self, sensitivity_variable, sensitivity_data_container: ContainerVariableDataHolderUnion, solution_step_index: int = 0):
+        key = f"problem_data/response_data/{sensitivity_data_container.GetModelPart().FullName()}/{self.GetName()}/sensitivities/{sensitivity_variable.Name()}/raw"
+        if solution_step_index == 0 and not self._optimization_info.HasValue(key):
```

Comment:
What would happen if the given key is wrong due to the wrong FullName, modelPart and etc. ?

Reply:
the key is an auto generated string, the model part name is taken from the corresponding container. So the model part name cannot be wrong. This is now called through the communicators so that, string manipulations are kept minimum to avoid developer mistakes.